### PR TITLE
GH-41121: [C++] Fix: left anti join filter empty rows.

### PR DESCRIPTION
**Pull request by Light-City**
### Rationale for this change

Since the left anti filter implementation is based on the left semi filter, and an assertion error occurs when the left semi filter rows are empty, this problem should be fixed.

### What changes are included in this PR?

swiss_join.cc and hash_join_node_test.cc

### Are these changes tested?
Yes

### Are there any user-facing changes?
No

* GitHub Issue: #41121

### CODE REVIEW

**Comment by Light-City** [comment]
I don’t know why lint ci hangs without any prompt?

**Comment by ZhangHuiGui** [comment]
> I don’t know why lint ci hangs without any prompt?

Seems you need to format your codes:
```c
 TEST(HashJoin, FilterEmptyRows) {
   BatchesWithSchema input_left;
-  input_left.batches = {ExecBatchFromJSON({int32(), utf8(), int32()}, R"([[2, "Jarry", 28]])")};
+  input_left.batches = {
+      ExecBatchFromJSON({int32(), utf8(), int32()}, R"([[2, "Jarry", 28]])")};
```

**Comment by Light-City** [comment]
> > I don’t know why lint ci hangs without any prompt?
> 
> Seems you need to format your codes:
> 
> ```c
>  TEST(HashJoin, FilterEmptyRows) {
>    BatchesWithSchema input_left;
> -  input_left.batches = {ExecBatchFromJSON({int32(), utf8(), int32()}, R"([[2, "Jarry", 28]])")};
> +  input_left.batches = {
> +      ExecBatchFromJSON({int32(), utf8(), int32()}, R"([[2, "Jarry", 28]])")};
> ```

thanks, I changed.

File: cpp/src/arrow/acero/swiss_join.cc
Comment (Light-City) [reply cpp/src/arrow/acero/swiss_join.cc:2163]: > It seems that `num_batch_rows` has been ensured to be `>0` in `GetNextBatch`, otherwise it should not enter the `FilterInner` logic.
> 
> https://github.com/apache/arrow/blob/831b94a65ebaaabe6fcad0ea378d0a784ef5bbb7/cpp/src/arrow/acero/swiss_join.cc#L1865
> 
> Have you missed some key issues?

This change is due to:
FilterLeftAnti calls FilterLeftSemi, and in my case the FilterOneBatch is a blank line. There is an assertion in probe_batch_builder.Flush() that requires the number of lines to be >0, so in order to solve this problem, the blank line needs to be determined here.

**Comment by Light-City** [comment]
test_table.py failed .The test failure has nothing to do with my PR, it seems to be asof join

File: cpp/src/arrow/acero/swiss_join.cc
Comment (zanmato1984) [inline cpp/src/arrow/acero/swiss_join.cc:2163]: Maybe it's better to move this shortcut below the line `*num_passing_rows = 0;` for the following two considerations:
1. The bunch of `ARROW_DCHECK` at the top of this function is doing basic logical checks thus we'd better do them all the time regardless of the shortcut.
2. This function doesn't assume the out parameter `num_passing_rows` has a proper initial value, so we should always assign it a meaningful one.

**Review by zanmato1984 — CHANGES_REQUESTED** [review]
This is a great catch, thank you!

One minor suggestion though.

**Comment by zanmato1984** [comment]
> test_table.py failed .The test failure has nothing to do with my PR, it seems to be asof join

Yeah, seems unrelated to this PR. Could be the same as https://github.com/apache/arrow/issues/41149.

**Review by zanmato1984 — APPROVED** [review]
+1

**Comment by zanmato1984** [comment]
We should wait for some committer to approve and merge. Thank you again for the fix!

cc @westonpace

**Comment by Light-City** [comment]
> We should wait for some committer to approve and merge. Thank you again for the fix!
> 
> cc @westonpace

Okay, thank you

**Comment by mapleFU** [comment]
@zanmato1984 should this better in 16.0?

**Comment by mapleFU** [comment]
I've change this patch to blocker

**Comment by pitrou** [comment]
Rebased on git main for CI.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

File: cpp/src/arrow/acero/hash_join_node_test.cc
Comment (pitrou) [inline cpp/src/arrow/acero/hash_join_node_test.cc:2040]:
```suggestion
  // Regression test for GH-41121.
  BatchesWithSchema input_left;
```

**Review by pitrou — APPROVED** [review]
This seems ok on the face of it, but I can't vouch for the correctness.

**Pull request merged**
